src/test/java/com/baidu/hugegraph/unit/concurrent/BarrierEventTest.java

```diff
@@ -68,38 +68,84 @@ public void testSignal() throws InterruptedException {
         int waitThreadNum = 10;
         ExecutorService executorService =
-                        Executors.newFixedThreadPool(waitThreadNum);
+                        Executors.newFixedThreadPool(waitThreadNum + 1);
         CountDownLatch waitLatch = new CountDownLatch(waitThreadNum);
         CountDownLatch signalLatch = new CountDownLatch(1);
         for (int i = 0; i < waitThreadNum; i++) {
             executorService.submit(() -> {
                 try {
-                    waitLatch.countDown();
+                    signalLatch.await();
                     barrierEvent.await();
-                    result.incrementAndGet();
+                    eventCount.incrementAndGet();
                 } catch (InterruptedException e) {
-                    // Do nothing
+                    waitThreadInterruptedCount.incrementAndGet();
+                } finally {
+                    waitLatch.countDown();
                 }
             });
         }
 
-        Thread signalThread = new Thread(() -> {
+        executorService.submit(() -> {
             barrierEvent.signal();
             signalLatch.countDown();
         });
-        signalThread.start();
+
+        executorService.shutdown();
+        executorService.awaitTermination(2L, TimeUnit.SECONDS);
         waitLatch.await();
+        Assert.assertEquals(10, eventCount.get());
+        Assert.assertEquals(0, waitThreadInterruptedCount.get());
+    }
+
+    @Test
+    public void testSignalByMultiThreadWithSignalLast()
+                throws InterruptedException {
+        BarrierEvent barrierEvent = new BarrierEvent();
+        AtomicInteger eventCount = new AtomicInteger(0);
+        AtomicInteger waitThreadInterruptedCount = new AtomicInteger(0);
+        AtomicInteger signalThreadInterruptedCount = new AtomicInteger(0);
+        int waitThreadNum = 10;
+        ExecutorService executorService =
+                        Executors.newFixedThreadPool(waitThreadNum + 1);
+        CountDownLatch waitLatch = new CountDownLatch(waitThreadNum);
+        CountDownLatch signalLatch = new CountDownLatch(1);
+        for (int i = 0; i < waitThreadNum; i++) {
+            executorService.submit(() -> {
+                try {
+                    waitLatch.countDown();
+                    barrierEvent.await();
+                    eventCount.incrementAndGet();
+                } catch (InterruptedException e) {
+                    waitThreadInterruptedCount.incrementAndGet();
+                }
+            });
+        }
+
+        executorService.submit(() -> {
+            try {
+                waitLatch.await();
+            } catch (InterruptedException e) {
+                signalThreadInterruptedCount.incrementAndGet();
+            }
+            barrierEvent.signal();
+            signalLatch.countDown();
+        });
         signalLatch.await();
-        TimeUnit.MICROSECONDS.sleep(100);
+        TimeUnit.MILLISECONDS.sleep(10L);
```

Comment:
must sleep?

Reply:
Yes, if reach this line, it indicates the signal thread has signaled, but the wait thread may not be notified and executed.

Reply:
please explain why instead of what,I mean seems we can remove the sleep